### PR TITLE
refactor(package-rules): delete all matchers from config 

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Remove all macther/excluders from the config after tha package rule has been merged in it.

## Context
These matchers should have been deleted as per the logic but it seems we missed them.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
I will add one more test which confirms that all matcher and excluder are deleted, so that we do not miss in the future.

File: lib/util/package-rules/index.ts [hunk 1]
```diff
@@ -91,6 +91,21 @@ export function applyPackageRules<T extends PackageRuleInputConfig>(
       delete config.matchCurrentValue;
       delete config.matchCurrentVersion;
       delete config.matchCurrentAge;
+      delete config.excludeDepNames;
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/util/package-rules/index.ts:94]: Yes. Trying that only

File: lib/util/package-rules/index.spec.ts [hunk 2]
```diff
@@ -43,6 +43,7 @@ describe('util/package-rules/index', () => {
       depName: 'a',
       isBump: true,
       currentValue: '1.0.0',
+      labels: ['bump'],
```
Comment (rarkins) [inline lib/util/package-rules/index.spec.ts:46]:
```suggestion
```

Change https://github.com/renovatebot/renovate/pull/26880/files#diff-b9fa5327dbf0d50ad7eb9d30c074f7cae334fc05bae58698e8b70206e9d29d14R60 instead to include the labels, this way you are validating that changes were applied

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/package-rules/index.ts [hunk 3]
```diff
@@ -73,25 +73,28 @@ export function applyPackageRules<T extends PackageRuleInputConfig>(
...
-      delete config.matchDepTypes;
-      delete config.matchCurrentValue;
-      delete config.matchCurrentVersion;
-      delete config.matchCurrentAge;
     }
   }
   return config;
 }
+
+function removeMatchers(
```
Comment (rarkins) [inline lib/util/package-rules/index.ts:89]: Is it better to create a new object like this or just to iterate through the existing one and delete applicable keys?

Comment (RahulGautamSingh) [reply lib/util/package-rules/index.ts:89]: I don't see any issue in both. It's a matter of choice since the performance difference isn't much.

Comment (rarkins) [reply lib/util/package-rules/index.ts:89]: In that case please refactor to delete

Comment (RahulGautamSingh) [reply lib/util/package-rules/index.ts:89]: Generally its bad practice to delete the keys while looping over an object, but it seems its okay in javasript.
[Reference ](https://stackoverflow.com/questions/3463048/is-it-safe-to-delete-an-object-property-while-iterating-over-them)

Comment (viceice) [reply lib/util/package-rules/index.ts:89]: Object.keys is evaluated before the loop, so no issue

Comment (rarkins) [reply lib/util/package-rules/index.ts:89]: It's already cloned due to mergeChildConfig

Comment (viceice) [reply lib/util/package-rules/index.ts:89]: ok, so delete is the only option here

File: lib/util/package-rules/index.ts [hunk 4]
```diff
@@ -73,25 +73,27 @@ export function applyPackageRules<T extends PackageRuleInputConfig>(
...
+  packageRule: PackageRule & PackageRuleInputConfig,
+): Record<string, unknown> {
+  for (const key of Object.keys(packageRule)) {
+    if (key.includes('match') || key.includes('exclude')) {
```
[the remark this answers is not in the document]
Comment (RahulGautamSingh) [reply lib/util/package-rules/index.ts:93]: There are some other options which can startwith exclude or match but they aren't the children of package-rules, so it is safe as any wrong configuration will have been caught during validation.

File: lib/util/package-rules/index.ts [hunk 5]
```diff
@@ -73,25 +73,27 @@ export function applyPackageRules<T extends PackageRuleInputConfig>(
...
     }
   }
   return config;
 }
+
+function removeMatchers(
+  packageRule: PackageRule & PackageRuleInputConfig,
+): Record<string, unknown> {
+  for (const key of Object.keys(packageRule)) {
+    if (key.startsWith('match') || key.startsWith('exclude')) {
```
Comment (viceice) [inline lib/util/package-rules/index.ts:93]: we should add a test which validates that every option starting with those prefixes are only allowed in package rules. so we don't silently delete wrong keys in future.

Comment (RahulGautamSingh) [reply lib/util/package-rules/index.ts:93]: Its already caught during config validation, where we check that the children is inside the correct parent or not.
https://github.com/renovatebot/renovate/blob/f69d7e83b33727cb66742facafb24e1c3ce1e432/lib/config/validation.ts#L235-L237

Comment (viceice) [reply lib/util/package-rules/index.ts:93]: that's not what I like to test.

I like to ensure nobody adds a `matchSomething` option to any other parent than package rules.

Comment (rarkins) [reply lib/util/package-rules/index.ts:93]: Well, they might be added to something like hostRules but then it's also safe to assume we can delete them from packageRules if found as they shouldn't be there in the first place and hopefully found during config validation

Comment (viceice) [reply lib/util/package-rules/index.ts:93]: ok, here is a concrete sample
- https://github.com/renovatebot/renovate/pull/26951/files#diff-0053f214ec552bf58e21e52b0af7fcfe1699c7be2989cee50c52d44c7046ed2fR2831

but you're right, this should not be here in that case

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**